packages/cli/src/UserManagement/PermissionChecker.ts

```diff
@@ -1,68 +1,76 @@
 import { Service } from 'typedi';
 import type { INode, Workflow } from 'n8n-workflow';
-import { CredentialAccessError, NodeOperationError, WorkflowOperationError } from 'n8n-workflow';
+import {
+	ApplicationError,
+	CredentialAccessError,
+	NodeOperationError,
+	WorkflowOperationError,
+} from 'n8n-workflow';
 
 import config from '@/config';
 import { License } from '@/License';
 import { OwnershipService } from '@/services/ownership.service';
 import { UserRepository } from '@db/repositories/user.repository';
 import { SharedCredentialsRepository } from '@db/repositories/sharedCredentials.repository';
-import { SharedWorkflowRepository } from '@db/repositories/sharedWorkflow.repository';
+import { ProjectService } from '@/services/project.service';
+import { RoleService } from '@/services/role.service';
 
 @Service()
 export class PermissionChecker {
 	constructor(
 		private readonly userRepository: UserRepository,
 		private readonly sharedCredentialsRepository: SharedCredentialsRepository,
-		private readonly sharedWorkflowRepository: SharedWorkflowRepository,
 		private readonly ownershipService: OwnershipService,
 		private readonly license: License,
+		private readonly projectService: ProjectService,
+		private readonly roleService: RoleService,
 	) {}
 
 	/**
 	 * Check if a user is permitted to execute a workflow.
+	 *
+	 * Allow:
+	 * - if user has global scope `workflow:execute`, OR
+	 * - if user has project scope `workflow:execute` in any of their roles
+	 * in any of the projects where the workflow is accessible, AND if every
+	 * credential used by nodes in the workflow is accessible by any of the
+	 * projects where the workflow is accessible.
 	 */
 	async check(workflowId: string, userId: string, nodes: INode[]) {
-		// allow if no nodes in this workflow use creds
-
-		const credIdsToNodes = this.mapCredIdsToNodes(nodes);
-
-		const workflowCredIds = Object.keys(credIdsToNodes);
-
-		if (workflowCredIds.length === 0) return;
-
-		// allow if requesting user is instance owner
-
-		const user = await this.userRepository.findOneOrFail({
-			where: { id: userId },
-		});
+		const user = await this.userRepository.findOneByOrFail({ id: userId });
```

Comment:
Would it be possible to remove `userId` and `user` entirely from the equation?

Since all executions should be part of saved workflows, the `userId` should not be needed anymore.

One caveat is that we can execute workflows form file or from JSON using the `Execute Workflow` node, but those should inherit the permissions from the parent workflow, so we should still be able to rely on `workflowId` above.

WDYT?